### ISSUE TITLE
TextFieldDriver doesn't export its tag if content is empty

### ISSUE THREAD

**Issue by hannan-azam** [OrchardCMS/Orchard#7493]
I reproduced this bug on 1.10.x
TextFieldDriver doesn't export its tag in recipe if content is empty. There is a check in Exporting function which exports the tag only if the value is not null or empty.

```
        protected override void Exporting(ContentPart part, TextField field, ExportContentContext context) {
            if (!String.IsNullOrEmpty(field.Value))
                context.Element(field.FieldDefinition.Name + "." + field.Name).SetAttributeValue("Text", field.Value);
        }
```

Because of this, when importing recipe on target application it skips the TextFieldDriver and doesn't replace the content with empty value.

That condition should be removed just like we have in BodyPartDriver

**Comment by sebastienros** [comment]
My instinct told me to look at the history for this file, and it appears we did this change on purpose here: https://github.com/hannan-azam/Orchard/commit/1edf94756c671b9c7040403154e5d99753498318

What should be the behavior then when we have this values:
- empty string
- null value
- some value (obvious)

/cc @DanielStolt

**Comment by DanielStolt** [comment]
The change I made was part of a bigger effort to make recipes leaner and more manageable, by having them not contain redundant information. I might have zeroed in too much on the scenario of **setup** recipes, where you use the recipe to initialize a new site. In this case, empty elements are useless and only serve to make the recipe bloated and the import slower.

However, in your case @hannan-azam you are using the recipe to import over an existing site, and you expect the empty value to be considered a value and overwrite any existing value with the empty one. I realize this is a perfectly reasonable use case too, which I did not consider.

I think the best solution is to make this exclusion conditional instead at the time of the export, by providing a new checkbox in the UI for the content step like so:

[ ] Exclude empty field values

If you check the box, empty values will not be exported. If you leave the box unchecked, everything will be exported. We must also remember to add support for this option to the command line interface (via the export configuration file support).

(I am aware that this changes the default from what it currently is, but I think it will lead to the most expected behavior for mest users that excluding empty values is the active opt-in choice.)

Regarding differentiating empty string from null values as @sebastienros suggests, I don't think that solves anything. In @hannan-azam's scenario you would want them included and processed in either case. I also vaguely recall that the difference between null and empty string is not very reliable in Orchard, depending on whether you have saved the item or not, etc. So in summary, I think we should continue to treat them as equivalents.

**Comment by sebastienros** [comment]
I would suggest to export all values for now, which would revert @DanielStolt 's changes, and in a second step add this option to exclude empty fields.
I think making the feature work correctly is more important than having the recipes look leaner.

How does that option flows to the filed drivers? I couldn't find any dynamic data structure containing the export screen values.

**Comment by DanielStolt** [comment]
@sebastienros Yeah there doesn't seem to be any information passed down today except for the content item itself. My suggestion would be to add either an overload or an optional parameter here:

```
XElement IContentManager.Export(ContentItem contentItem, ExportContentOptions options = null)
```

The `ExportContentOptions` would at this point only contain `bool ExcludeEmptyFields` and maybe other things in the future. These options could then be added to the `ExportContentContext` object which in turn gets passed to all the drivers.

If you agree I can submit a PR.

**Comment by DanielStolt** [comment]
Revert is now merged in ac978b665b81b1e0d70594af53b94d6d699a7955.

I will keep the issue assigned to myself and implement the second part in dev.
